src/security/PeerConnector.cc

```diff
@@ -679,11 +759,23 @@ Security::PeerConnector::certDownloadingDone(SBuf &obj, int downloadStatus)
         char buffer[1024];
         debugs(81, 5, "Retrieved certificate: " << X509_NAME_oneline(X509_get_subject_name(cert), buffer, 1024));
         ContextPointer ctx(getTlsContext());
-        const Security::CertList &certsList = srvBio->serverCertificatesIfAny();
-        if (const char *issuerUri = Ssl::uriOfIssuerIfMissing(cert, certsList, ctx)) {
-            urlsOfMissingCerts.push(SBuf(issuerUri));
+        const auto certsList = SSL_get_peer_cert_chain(&sconn);
+        if (Ssl::issuerIsMissing(cert, certsList, ctx)) {
+            if (const auto issuerUri = Ssl::hasAuthorityInfoAccessCaIssuers(cert))
+                urlsOfMissingCerts.push(SBuf(issuerUri));
+            else {
+                debugs(81, 5, "Can not retrieve issuer for certificate " << X509_NAME_oneline(X509_get_subject_name(cert), buffer, 1024) << " abort");
```

Comment:
Well.  Imagine the following scenario:
 - the server send 3 certificates
 -  the issuer certificate is missing  from 1st, and 3rd certificates.
 -  we can download missing issuer certificate only for the 1st certificate the 3rd lacks the downloading info. 
 - only the 1st and 2nd really needed for server validation. Extra certificates breaks protocol, however  a missing certificate also breaks protocol if I am not wrong. I know that OpenSSL (at least some releases) does not check for such errors, the openSSL validation will succeed in this case.

Does this scenario looks reasonable?

> If there is no firm guarantee that the validation will now fail, then we cannot short-circuit; we should continue fetching urlsOfMissingCerts.

I think it is better to continue downloading if we can. In the above scenario maybe there are 4 certificates where the 4th is the issuer certificate of 2nd's issuer certificate and has a missing certificate itself.
In practice I do not believe that we could hit a such case however.

To fix squid continue downloading I think requires to fix the missingChainCertificatesUrls function and related code.

>  If there is a firm guarantee that the validation will now fail, then perhaps we should throw an error here instead of waiting for the our validation to fail some time later

I think this is wrong in any case. If the validation passes then all are ok. If validation fails will produce also a good error page for the user.

Reply:
> I think it is better to continue downloading if we can. ... To fix squid continue downloading I think requires to fix the missingChainCertificatesUrls function and related code.

Sounds good to me. Let's keep download until there is hope that downloading eventually completes the chain. Please change the code accordingly. There should be a corresponding source code comment to note/justify this non-obvious strategy.

I have not checked all the details, but I suspect that missingChainCertificatesUrls() adjustments would be minimal -- it should return true if its first parameter (i.e. urlsOfMissingCerts) is not empty, even if we do not know how to download some of the other missing certificates.

> > perhaps we should throw here

> I think this is wrong in any case. If the validation passes then all are ok. If validation fails will produce also a good error page for the user.

Well, the throwing code can relay enough details to produce a good error page. I do not think that is technically impossible -- we know that there is a problem and can report it somehow. However, I agree that it would be easier to use the existing error handling code for that.

Reply:
commit c00e8b017

> There should be a corresponding source code comment to note/justify this non-obvious strategy.

The Ssl::missingChainCertificatesUrls documents that it is now return if at least one URI found and I add a comment inside Security::PeerConnector::certDownloadingDone.
I hope it is OK.